src/test/epoll/test_epoll_edge.rs

```diff
@@ -0,0 +1,414 @@
+    let mut events = Vec::new();
+    events.resize(10, epoll::EpollEvent::empty());
+
+    let t0 = std::time::Instant::now();
+
+    let res = epoll::epoll_wait(
+        epoll_fd,
+        &mut events,
+        timeout.as_millis().try_into().unwrap(),
+    );
+
+    let t1 = std::time::Instant::now();
+
+    events.resize(res.unwrap_or(0), epoll::EpollEvent::empty());
+
+    if do_read {
+        for ev in &events {
+            let fd = ev.data() as i32;
+            // we don't care if the read is successful or not (another thread may have already read)
+            let _ = unistd::read(fd, &mut [0]);
+        }
+    }
+
+    WaiterResult {
+        duration: t1.duration_since(t0),
+        epoll_res: res,
+        events,
+    }
+}
+
+fn test_multi_write(readfd: libc::c_int, writefd: libc::c_int) -> anyhow::Result<()> {
+    let epollfd = epoll::epoll_create()?;
+
+    test_utils::run_and_close_fds(&[epollfd, readfd, writefd], || {
+        let mut event = epoll::EpollEvent::new(EpollFlags::EPOLLET | EpollFlags::EPOLLIN, 0);
+        epoll::epoll_ctl(
+            epollfd,
+            epoll::EpollOp::EpollCtlAdd,
+            readfd,
+            Some(&mut event),
+        )?;
+
+        let timeout = Duration::from_millis(100);
+
+        let thread = std::thread::spawn(move || {
+            vec![
+                do_epoll_wait(epollfd, timeout, /* do_read= */ false),
+                do_epoll_wait(epollfd, timeout, /* do_read= */ false),
+                // The last one is supposed to timeout.
+                do_epoll_wait(epollfd, timeout, /* do_read= */ false),
+            ]
+        });
+
+        // Wait for readers to block.
+        std::thread::sleep(timeout / 3);
+
+        // Make the read-end readable.
+        unistd::write(writefd, &[0])?;
+
+        // Wait again and make the read-end readable again.
+        std::thread::sleep(timeout / 3);
+        unistd::write(writefd, &[0])?;
+
+        let results = thread.join().unwrap();
+
+        // The first two waits should have received the event
+        for res in &results[..2] {
+            ensure_ord!(res.epoll_res, ==, Ok(1));
+            ensure_ord!(res.duration, <, timeout);
+            ensure_ord!(res.events[0], ==, epoll::EpollEvent::new(EpollFlags::EPOLLIN, 0));
+        }
+
+        // The last wait should have timed out with no events received.
+        ensure_ord!(results[2].epoll_res, ==, Ok(0));
+        ensure_ord!(results[2].duration, >=, timeout);
+
+        Ok(())
+    })
+}
+
+fn test_write_then_read(readfd: libc::c_int, writefd: libc::c_int) -> anyhow::Result<()> {
+    let epollfd = epoll::epoll_create()?;
+
+    test_utils::run_and_close_fds(&[epollfd, readfd, writefd], || {
+        let mut event = epoll::EpollEvent::new(EpollFlags::EPOLLET | EpollFlags::EPOLLIN, 0);
+        epoll::epoll_ctl(
+            epollfd,
+            epoll::EpollOp::EpollCtlAdd,
+            readfd,
+            Some(&mut event),
+        )?;
```

Comment:
Can you rename this to `test_write_then_partial_read`, and then remove the test with the same name from `test_epoll.rs`?